lock-api/src/main/java/com/palantir/lock/watch/NoOpLockWatchEventCache.java

```diff
@@ -58,11 +67,34 @@ public TransactionsLockWatchUpdate getUpdateForTransactions(
             Set<Long> startTimestamps, Optional<LockWatchVersion> version) {
         return ImmutableTransactionsLockWatchUpdate.builder()
                 .clearCache(true)
-                .startTsToSequence(
-                        startTimestamps.stream().collect(Collectors.toMap(startTs -> startTs, $ -> FAKE_VERSION)))
+                .startTsToSequence(startTimestamps.stream()
+                        .collect(Collectors.toMap(startTs -> startTs, $ -> currentVersion.orElse(FAKE_VERSION))))
                 .build();
     }
 
     @Override
     public void removeTransactionStateFromCache(long startTimestamp) {}
+
+    private void updateVersion(Optional<LockWatchVersion> maybeNewVersion) {
+        currentVersion = maybeNewVersion.map(newVersion -> currentVersion
+                .filter(current -> current.id().equals(newVersion.id()) && current.version() > newVersion.version())
+                .orElse(newVersion));
+    }
+
+    private Optional<LockWatchVersion> extractVersionFromUpdate(LockWatchStateUpdate update) {
```

Comment:
Unclear if this has been a convention throughout Lock Watches, but the decomp - in particular, rejecting a success from post-leader election without a snapshot _here_ - seems a bit unexpected. I'd prefer this just return the version, and updateVersion take care of that filtering. It looks like you do that anyway in updateVersion!

Reply:
So in the real implementation, a success after a leader election _will_ fail (as in, the version will be empty). `updateVersion` doesn't quite do the same thing - if there has been a leader election without a snapshot, `updateVersion` will set it to empty, whereas if we permitted this, `currentVersion` would have a value.

Reply:
The test `leaderChangeClearsCache` in `LWECIT` demonstrates this behaviour - accepting a success from a post-leader election causes the versions to diverge.

Reply:
There's a good reason for this - we can't have a valid version unless we've had a snapshot telling us:
a) what descriptors are being watched;
b) what the currently held watches are.

Otherwise we could tell the client that there are only X watches taken out, when that may not be accurate.

Reply:
Ah, I misread the updateVersion code - I've left a comment there.

I still find `extractVersionFromUpdate` to return an Optional weird: the update clearly _has_ a log ID and a version, just it may not be one we're interested in. I think we could

1. add a brief comment indicating the properties of when this returns the version
2. name this `extractRelevantVersion...`?

I don't have a good name for 2 so suggest 1.